Add docs and blog post for ESC GitHub Action (#14268)
pulumi/docs: content/blog/announcing-pulumi-esc-github-action/index.md

---

> title: "Announcing the Pulumi ESC GitHub Action"

Reviewer: Does the action not have a "migrate GH secrets to ESC" sub action?

Author: No, it does not. I can add the bash script I used to build it in the docs. We can leave building that into the action as a future improvement.

date: "2025-03-05"
meta_desc: "Simplify Secrets Management and CI/CD with Pulumi ESC."
authors: 
  - "komal-ali"
tags:
  - esc
  - secrets
  - github
  - features
---

We’re excited to share our latest addition to the Pulumi Ecosystem: the [Pulumi ESC GitHub Action](https://github.com/marketplace/actions/esc-action). This new GitHub Action streamlines how you manage secrets, configurations, and CI/CD tasks using [Pulumi ESC](/product/esc), helping your team securely scale and automate your software delivery pipelines.

> ---more

Reviewer 2: ```suggestion
```

> ## Why We Built It

Reviewer 2: ```suggestion
## Why we built it
```

Reviewer 2: Would you be able to fix the rest of the headings to use sentence casing, which we're trying to be consistent with. https://github.com/pulumi/docs/blob/master/STYLE-GUIDE.md. - thx!!

> Secrets management continues to be a challenge for organizations of all sizes. Storing long-lived credentials in multiple places (e.g., GitHub, third-party CI systems, and internal servers) often leads to “secret sprawl” and potential security gaps. With Pulumi ESC, you can centralize and manage short-lived dynamic credentials, rotate them automatically, and ensure secrets don’t persist beyond their necessary lifetime.

Reviewer: ```suggestion
Secrets management continues to be a challenge for organizations of all sizes. Storing long-lived credentials in multiple places (e.g., GitHub, third-party CI systems, and internal servers) often leads to “secret sprawl” and potential security gaps. With Pulumi ESC, you can migrate to dynamic credentials, and ensure secrets don’t persist beyond their necessary lifetime.  If your process requires longer-lived static credentials, with ESC you can manage credentials in a central place and rotate them automatically.
```

We should have a link to secrets rotation blog post or docs.

Author: It's linked later in the doc, but I can link again here.

By packaging ESC functionality into a GitHub Action, we’re enabling teams to inject secrets and run ESC commands directly within their GitHub workflows. This allows for a streamlined approach to environment management and secrets rotation—without leaving GitHub.

> ## Key Benefits

Reviewer 2: ```suggestion
## Key benefits
```

2.	**Automatic Secret Rotation** - Pulumi ESC handles rotation for you, ensuring that any secrets or credentials you inject into GitHub Actions remain fresh and invalid once your workflow completes.
3.	**Streamlined CI/CD Pipeline** - In addition to secrets management, you can run any Pulumi ESC command directly in your GitHub Actions workflows. Automate the creation, update, or teardown of environments as part of your CI/CD process.
4.	**Seamless Integration with Pulumi Cloud** - Combine the Pulumi ESC GitHub Action with [pulumi/auth-actions](https://github.com/marketplace/actions/pulumi-auth-action) to authenticate securely with Pulumi Cloud. No need for long-lived tokens—simply leverage short-lived tokens provided at runtime.

## How It Works

At a high level, the Pulumi ESC GitHub Action can do three things:

1. **Download the Pulumi ESC CLI** - By default, the action ensures the ESC CLI is installed in your GitHub Actions environment. If you provide a specific version, that version is installed.

> 2. **Inject all environment variables from an ESC environment** - If you specify an environment (for example, tinyco/someProject/myEnv), the action will inject all environment variables and projected files associated with that environment into the GitHub Actions context.

Reviewer 2: ```suggestion
2. **Inject all environment variables from an ESC environment** - If you specify an environment (for example, `tinyco/someProject/myEnv`), the action will inject all environment variables and projected files associated with that environment into the GitHub Actions context.
```

3. **Inject specific environment variables from an ESC environment** - If you only need certain keys, you can specify them in the action’s inputs to pick and choose what gets injected, giving you fine-grained control.

## Example Workflow

Below is a minimal workflow that showcases injecting an entire environment’s variables into your GitHub Actions job:

```yaml
on:
  - pull_request

permissions:
  id-token: write
  contents: read

jobs:
  test-env-injection:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repository
        uses: actions/checkout@v4
      - name: Authenticate with Pulumi Cloud
        uses: pulumi/auth-actions@v1
        with:
          organization: pulumi
          requested-token-type: urn:pulumi:token-type:access_token:organization
      - name: Install and inject ESC environment variables
        uses: pulumi/esc-action@v1
        with:

>           environment: 'tinyco/someProject/myEnv'

Reviewer: Can I pull in a specific version tag? Should we show that?

---
title: "Announcing the Pulumi ESC GitHub Action"

> date: "2025-03-04"

Reviewer 3: ```suggestion
date: "2025-03-05"
```

> meta_desc: "Simplify Secrets Management and CI/CD with Pulumi ESC."

Reviewer 3: ```suggestion
meta_desc: "Simplify CI/CD Secrets and Management with Pulumi ESC."
```

Author: Hmm... the suggested change doesn't really make sense to me.

authors: 
  - "komal-ali"
tags:
  - esc
  - secrets
  - github

> ---

Reviewer 3: Add `features` tag

> We’re excited to share our latest addition to the Pulumi Ecosystem: the [Pulumi ESC GitHub Action](https://github.com/marketplace/actions/esc-action). This new GitHub Action streamlines how you manage secrets, configurations, and CI/CD tasks using Pulumi ESC, helping your team securely scale and automate your software delivery pipelines.

Reviewer 3: ```suggestion
We’re excited to share our latest addition to the Pulumi Ecosystem: the [Pulumi ESC GitHub Action](https://github.com/marketplace/actions/esc-action). This new GitHub Action streamlines how you manage secrets, configurations, and CI/CD tasks using [Pulumi ESC](/product/esc), helping your team securely scale and automate your software delivery pipelines.
```

> 

Reviewer 3: Add "--- more" here so that Why we built it doesn't show up in the Blog list page.

## Why We Built It

> Secrets management continues to be a challenge for organizations of all sizes. Storing long-lived credentials in multiple places (e.g., GitHub, third-party CI systems, and internal servers) often leads to “secret sprawl” and potential security gaps. With Pulumi ESC, you can centralize and manage short-lived dynamic credentials, rotate them automatically, and ensure secrets don’t persist beyond their necessary lifetime.

Reviewer 3: We could add an example here of how we often see organizations storing long-lived cloud provider credentials in their CI/CD pipelines

Author: Wanna draft some words for me I can pop in there?

2.	**Automatic Secret Rotation** - Pulumi ESC handles rotation for you, ensuring that any secrets or credentials you inject into GitHub Actions remain fresh and invalid once your workflow completes.
3.	**Streamlined CI/CD Pipeline** - In addition to secrets management, you can run any Pulumi ESC command directly in your GitHub Actions workflows. Automate the creation, update, or teardown of environments as part of your CI/CD process.
4.	**Seamless Integration with Pulumi Cloud** - Combine the Pulumi ESC GitHub Action with [pulumi/auth-actions](https://github.com/marketplace/actions/pulumi-auth-action) to authenticate securely with Pulumi Cloud. No need for long-lived tokens—simply leverage short-lived tokens provided at runtime.

## How It Works

At a high level, the Pulumi ESC GitHub Action can do three things:

1. **Download the Pulumi ESC CLI** - By default, the action ensures the ESC CLI is installed in your GitHub Actions environment. If you provide a specific version, that version is installed.

> 2. **Inject all environment variables from an ESC environment** - If you specify an environment (for example, tinyco/someProject/myEnv), the action will inject all environment variables and projected files associated with that environment into the GitHub Actions context.

Reviewer 3: Is it only environment variables, or can I reference secrets/config in the values section? And is there a scenario where we would do that?

Author: It's only environment variables for now. There are complications with other secrets/config since they can be complex structures.

3. **Inject specific environment variables from an ESC environment** - If you only need certain keys, you can specify them in the action’s inputs to pick and choose what gets injected, giving you fine-grained control.

## Example Workflow

Below is a minimal workflow that showcases injecting an entire environment’s variables into your GitHub Actions job:

```yaml
on:
  - pull_request

permissions:
  id-token: write
  contents: read

jobs:
  test-env-injection:
    runs-on: ubuntu-latest
    steps:
      - name: Check out repository
        uses: actions/checkout@v4
      - name: Authenticate with Pulumi Cloud
        uses: pulumi/auth-actions@v1
        with:
          organization: pulumi
          requested-token-type: urn:pulumi:token-type:access_token:organization
      - name: Install and inject ESC environment variables
        uses: pulumi/esc-action@v1
        with:
          environment: 'tinyco/someProject/myEnv'
      - name: Verify environment variables
        run: |
          echo "FOO=$FOO"
          echo "SOME_IMPORTANT_KEY=$SOME_IMPORTANT_KEY"
          echo "TEST_ENV=$TEST_ENV"
```

> ## What This Does

Reviewer 3: ```suggestion
The GitHub Action workflow above does the following: 
```

1.	Authenticate with Pulumi Cloud using pulumi/auth-actions. Instead of storing a permanent token, we use a short-lived token to authenticate with Pulumi Cloud.
2.	Inject the environment variables from tinyco/someProject/myEnv into the current GitHub Actions environment, making them accessible to subsequent steps in the workflow.

## Use Cases

1.	**Dynamic Credentials in CI/CD** - Inject ephemeral secrets for short-lived tasks, such as integrating with third-party APIs or standing up ephemeral infrastructure. This ensures credentials are invalidated once your pipeline finishes running.
2.	**Automated Environment Management** - Use Pulumi ESC commands to create and manage environments directly from GitHub Actions. This can be part of a larger GitOps strategy where infrastructure changes are tracked in version control.
3.	**Secret Rotation for Compliance** - By using ESC’s built-in rotation, you can achieve compliance requirements and best practices without manual overhead or reliance on multiple external secrets stores.

> ## Getting Started

Reviewer 3: Getting started seems redundant since you already have an example.

1.	**Add the Pulumi ESC Action** - In your existing GitHub Actions workflow, reference pulumi/esc-action@v1.
2.	**Set Up Pulumi ESC** - Ensure you have your environment defined in Pulumi ESC with the required variables and files.
3.	**Authenticate Securely with Pulumi Cloud** - We recommend using pulumi/auth-actions to generate and pass a short-lived token as PULUMI_ACCESS_TOKEN.
4.	**Run and Validate** - Push your changes to GitHub, run your workflow, and verify that your secrets and environment variables are successfully injected.

## What’s Next?

> - Explore More ESC Features: Dive deeper into ESC’s environment management, secrets rotation, and dynamic credentials.

Reviewer 3: Add links to 'environment management', secrets rotation, dynamic credentials. 

Link to GitHub Action open source code
